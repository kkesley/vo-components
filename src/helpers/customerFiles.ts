import {
  CustomerFile,
  CustomerFileRequestType,
  CollectionAssetType,
  CustomerFileType,
  CustomerFileTypeSystemIdentifier,
} from 'graphql/types'

export interface FlatCustomerFile {
  bucket?: string
  file_path: string
  file_name: string | undefined
  file_id: string
  request_type: CustomerFileRequestType
  request_timestamp: string
  requester?: string
  customer_id: string
  collection_id?: string
  collection_segment?: CollectionAssetType
  is_ready?: boolean
  type?: CustomerFileType
  system_identifier?: CustomerFileTypeSystemIdentifier
}

export function flattenCustomerFiles(
  customerFiles: CustomerFile[]
): FlatCustomerFile[] {
  return (
    customerFiles.flatMap<FlatCustomerFile>(
      (customerFile) =>
        customerFile.output_result?.target_paths?.map((path) => ({
          bucket: customerFile.output_result?.target_bucket || undefined,
          file_path: path,
          file_name: path.split('/').pop(),
          file_id: customerFile.file_id,
          request_type: customerFile.request_type,
          request_timestamp: customerFile.request_timestamp,
          requester: customerFile.requester || undefined,
          customer_id: customerFile.customer_id,
          collection_id: customerFile.collection_id || undefined,
          collection_segment: customerFile.collection_segment || undefined,
          is_ready: customerFile.is_ready || undefined,
          type: customerFile.type || undefined,
          system_identifier: customerFile.system_identifier || undefined,
        })) || []
    ) || []
  )
}

export function getFileExtension(file_path: string): string | undefined {
  return file_path.toUpperCase().split('.').pop()
}
