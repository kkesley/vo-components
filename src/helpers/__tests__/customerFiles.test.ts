import { flattenCustomerFiles, getFileExtension } from '../customerFiles'
import {
  CustomerFile,
  CustomerFileRequestType,
  CollectionAssetType,
  CustomerFileType,
  CustomerFileTypeSystemIdentifier,
} from '../../graphql/types'

describe('flattenCustomerFiles', () => {
  const customerFiles: CustomerFile[] = [
    {
      file_id: 'file-a',
      request_type: CustomerFileRequestType.Zip,
      request_timestamp: '2020-01-01 02:02',
      requester: 'kendrick',
      customer_id: 'customer-a',
      collection_id: 'collection-a',
      collection_segment: CollectionAssetType.Image,
      is_ready: true,
      output_result: {
        target_bucket: 'some-bucket',
        target_paths: [
          'protected/customer-a/misc/abc.jpg',
          'protected/customer-a/misc/cde.zip',
        ],
      },
      input_option: {
        original_paths: ['abc.jpg'],
      },
      type: CustomerFileType.SystemGenerated,
      system_identifier: CustomerFileTypeSystemIdentifier.ZipAll,
    },
    {
      file_id: 'file-b',
      request_type: CustomerFileRequestType.Zip,
      request_timestamp: '2020-01-01 02:02',
      requester: 'veronica',
      customer_id: 'customer-b',
      collection_id: 'collection-b',
      collection_segment: CollectionAssetType.ImageRaw,
      is_ready: true,
      output_result: {
        target_paths: ['protected/customer-b/misc/pop.zip'],
      },
      input_option: {
        original_paths: ['abc.jpg'],
      },
      type: CustomerFileType.UserDefined,
    },
    {
      file_id: 'file-c',
      request_type: CustomerFileRequestType.Zip,
      request_timestamp: '2020-01-01 02:02',
      requester: 'kendrick again',
      customer_id: 'customer-c',
      collection_id: 'collection-c',
      collection_segment: CollectionAssetType.ImageRawSelected,
      is_ready: true,
      output_result: {
        target_paths: [],
      },
      input_option: {
        original_paths: ['abc.jpg'],
      },
      type: CustomerFileType.UserDefined,
    },
  ]
  it('retuns the correct customer files', () => {
    expect(flattenCustomerFiles(customerFiles)).toMatchSnapshot()
  })
})

describe('getFileExtension', () => {
  const testCases = [
    {
      path: 'a/b/c/abc.jpg',
      extension: 'JPG',
    },
    {
      path: 'a/b/c/abc.zip',
      extension: 'ZIP',
    },
    {
      path: 'abc.ZiP',
      extension: 'ZIP',
    },
  ]
  testCases.forEach((testCase) => {
    it(`returns the correct extension for ${testCase.path}`, () => {
      expect(getFileExtension(testCase.path)).toEqual(testCase.extension)
    })
  })
})
