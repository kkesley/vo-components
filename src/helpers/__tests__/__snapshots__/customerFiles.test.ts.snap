// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flattenCustomerFiles retuns the correct customer files 1`] = `
Array [
  Object {
    "bucket": "some-bucket",
    "collection_id": "collection-a",
    "collection_segment": "IMAGE",
    "customer_id": "customer-a",
    "file_id": "file-a",
    "file_name": "abc.jpg",
    "file_path": "protected/customer-a/misc/abc.jpg",
    "is_ready": true,
    "request_timestamp": "2020-01-01 02:02",
    "request_type": "ZIP",
    "requester": "kendrick",
    "system_identifier": "ZIP_ALL",
    "type": "SYSTEM_GENERATED",
  },
  Object {
    "bucket": "some-bucket",
    "collection_id": "collection-a",
    "collection_segment": "IMAGE",
    "customer_id": "customer-a",
    "file_id": "file-a",
    "file_name": "cde.zip",
    "file_path": "protected/customer-a/misc/cde.zip",
    "is_ready": true,
    "request_timestamp": "2020-01-01 02:02",
    "request_type": "ZIP",
    "requester": "kendrick",
    "system_identifier": "ZIP_ALL",
    "type": "SYSTEM_GENERATED",
  },
  Object {
    "bucket": undefined,
    "collection_id": "collection-b",
    "collection_segment": "IMAGE_RAW",
    "customer_id": "customer-b",
    "file_id": "file-b",
    "file_name": "pop.zip",
    "file_path": "protected/customer-b/misc/pop.zip",
    "is_ready": true,
    "request_timestamp": "2020-01-01 02:02",
    "request_type": "ZIP",
    "requester": "veronica",
    "system_identifier": undefined,
    "type": "USER_DEFINED",
  },
]
`;
